Rework FluidProp error message to be more indicative

GiudGiud opened the pull request:
refs #27207

GiudGiud committed:
Rework error message to be more indicative
refs #27207

loganharbour commented:
FYI including name() here is redundant; the class and object name will show up in the error prefix

GiudGiud commented:
I removed name.
Type is kind of part of the message. I think it s fine for an error

GiudGiud commented:
Test failures unrelated

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2145317/) on cbf0cff wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27208/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2145332/) on cbf0cff wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>225684</th>
      <th colspan=4>#27208 cbf0cf</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.29%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>101438</td>
      <td>101437</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17500</td>
      <td>17501</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27208/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27208/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27208/coverage/xfem)

</details>



This comment will be updated on new commits.

joshuahansel commented on modules/fluid_properties/include/fluidproperties/SinglePhaseFluidProperties.h:
```diff
@@ -39,7 +39,14 @@
-    fluidPropError(name(), ": ", __PRETTY_FUNCTION__, " derivative derivatives not implemented."); \
+    fluidPropError("The fluid properties class '",                                                 \
+                   type(),                                                                         \
+                   "' has not implemented the derivatives for the automatic "                      \
+                   "differentiation-enabled method below. You can choose to neglect the "          \
+                   "derivatives by passing the 'allow_incorrect_jacobian' parameter. If your "     \
+                   "application requires this method, you "                                        \
+                   "must either implement it or use a different fluid properties class.\n\n",      \
```
```suggestion
                   "' has not implemented the automatic "                      \
                   "differentiation-enabled method below. You can choose to neglect the "          \
                   "derivatives by setting the 'allow_imperfect_jacobians' parameter. If your "     \
                   "application requires this method, you "                                        \
                   "must either implement it or use a different fluid properties class.\n\n",      \
```

joshuahansel replied:
However, I think a better design would be to move that bit about setting that parameter into `fluidPropError`, but only into the error message, not the warning, since for the latter, they've already set that parameter. Also bonus points (and I can do this as follow-up): `fluidPropError` is a vague name that says nothing that's it about the derivatives not being implemented, so I suggest we replace `fluidPropError(args)` with `unimplementedDerivativeMethod(pretty_name)` and `unimplementedNonderivativeMethod(pretty_name)`. Also I'd also add in the warning a note about the derivatives being approximate now.

GiudGiud replied:
ok I did that. You're not wrong for the clarity. I think the idea was to use the same method / warning everywhere for these errors but it only ever got used for derivatives (in the source of the module at least. Maybe I m going to hit relap7 errors next. We'll see)

joshuahansel commented on modules/fluid_properties/include/fluidproperties/SinglePhaseFluidProperties.h:
```diff
@@ -54,13 +61,24 @@
-    fluidPropError(name(), ": ", __PRETTY_FUNCTION__, " derivatives not implemented.");            \
+    fluidPropError("The fluid properties class '",                                                 \
+                   type(),                                                                         \
+                   "' has not implemented the derivatives for the method below. You can choose "   \
+                   "to neglect the derivatives by passing the 'allow_incorrect_jacobian' "         \
+                   "parameter. If your application requires this method, you must either "         \
+                   "implement it or use a different fluid properties class.\n\n",                  \
```
```suggestion
                   "' has not implemented the method below. You can choose "   \
                   "to neglect the derivatives by setting the 'allow_imperfect_jacobians' "         \
                   "parameter. If your application requires this method, you must either "         \
                   "implement it or use a different fluid properties class.\n\n",                  \
```

joshuahansel reviewed (changes requested):
As a side note, `fluidPropError` is kinda weird

GiudGiud committed:
Address Josh review
- dont offer to neglect derivatives if already neglecting
- rename method and de-template it, making it a lot simpler

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2151490/) on 7851e46 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27208/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27208/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ad6651e155461479cbb4aca2a24c1a784fb1edc4`
